### PR TITLE
Add support for pipeline parameters

### PR DESCRIPTION
**Pull request by kiersten-stokes**
This PR, once out of draft, will address all bullet points in #2994 with the exception of the CLI support.

Related: discussion #2970

### What changes were proposed in this pull request?
This PR includes both the frontend and backend changes required to support pipeline parameters. 

- a new API endpoint and handler to fetch pipeline parameter schema `/pipeline/{processor_type}/parameters`
- a new class to represent a well-defined `PipelineParameter` object (one for each runtime type that supports parameters) and supply the schema with runtime-specific input types
- add plumbing to support a fourth "input type" (aka "widget type") called `parameter` for relevant custom component properties
- add an extra property `pipeline_parameters` to the generic component properties
- add support for converting and passing parameters to the `Pipeline` object after submission
- add support for adding pipeline parameters to the generated DSL
- add support for passing pipeline parameters as arguments to the KFP bootstrapper and setting them as environment variables in the bootstrapped container

NOTE: After some discussion, it was decided that generic pipelines should support the set of properties/parameters/etc that are supported by **all** runtimes (ie an intersection). While changes will be required to make this true of how properties are handled, this pipeline parameter support PR does follow this directive.

### TODO

- [x] **backend**: rename variables and files that include the substring `parameter` to use `property` instead (where applicable)
- [x] **backend**: implement processing and validation logic for custom and generic components
- [x] **backend**: fix KFP icon rendering
- [x] **backend**: add backend tests
- [x] **backend**: address all `TODO`s
- [x] **backend**: add validation and/or trap error during `compile` for mismatched KFP types
- [x] **backend/frontend**: add description to individual fields for each parameter
- [x] **backend/frontend**: only allow selection of parameter type(s) that match the given component-defined type - may be better in a follow-up PR
- [x] **frontend**: add support for new third "Pipeline Parameters" panel tab
- [x] **frontend**: fix style for tooltip
- [x] **frontend**: fix validation issues
- [x] **frontend**: add types to submit/export dialog
- [x] **frontend**: (nice to have) display custom message in above tab for runtimes that do not support parameters
- [x] **frontend**: add params to the appropriate location in the pipeline JSON
- [x] **frontend**: add new widget type named `parameter`; component properties will appear as below in the pipeline JSON
  - ```json "url": {
    "url": {
      "widget": "parameter",
      "value": "user_defined_param_1"
    },
- [x] **frontend**: add ability to change pipeline parameter default values on pipeline submit
- [x] **frontend**: add ability to change pipeline parameter default values on pipeline export
- [x] **frontend**: implement parameter widget selection for custom components
- [x] **frontend**: implement parameter property for generic components
- [x] **documentation**: add pipeline parameters content to docs or open another issue/PR to address -> #3046 

### How was this pull request tested?
Manual testing.
Added new backend tests.

NOTE: To test this PR as it relates to generic components, you will have to update the `ELYRA_BOOTSTRAP_SCRIPT_URL` env var to point to this branch `https://raw.githubusercontent.com/kiersten-stokes/elyra/pipeline-params/elyra/kfp/bootstrapper.py`
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: elyra/pipeline/properties.py [hunk 1]
```diff
@@ -49,41 +51,78 @@
...
+        **kwargs,
+    ):
+        # Get type info from passed in type dict, if available; default to standard python types
+        allowed_input_types = kwargs.get("allowed_input_types")
+        if not allowed_input_types:
+            allowed_input_types = self._python_input_types
+
+        if base_type not in allowed_input_types:
+            raise ValueError(
+                f"Invalid property type '{base_type}': valid types are f{list(allowed_input_types.keys())}."
```
Comment (ptitzler) [inline elyra/pipeline/properties.py:84]: No f?

```suggestion
                f"Invalid property type '{base_type}': valid types are {list(allowed_input_types.keys())}."
```

Comment (kiersten-stokes) [reply elyra/pipeline/properties.py:84]: good catch!

File: elyra/pipeline/kfp/kfp_properties.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                KfpPropertyInputType(base_type="Float"),
+                KfpPropertyInputType(base_type="Bool"),
+            ],
+            hidden=True,
+            required=False,
+            use_in_key=False,
+        ),
+        ListItemPropertyAttribute(
+            attribute_id="required",
+            description="Whether a value is required for this property during pipeline submit/export",
```
Comment (ptitzler) [inline elyra/pipeline/kfp/kfp_properties.py:122]:
```suggestion
            description="Whether a value is required for this parameter during pipeline submit/export",
```

File: elyra/pipeline/kfp/kfp_properties.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+
+class KfpPipelineParameter(PipelineParameter):
+    """An ElyraProperty representing a single pipeline parameter for the Kubeflow Pipelines runtime"""
+
+    property_id = "KFP_PIPELINE_PARAMETERS"
+    property_attributes = [
+        ListItemPropertyAttribute(
+            attribute_id="name",
+            description="The name of the parameter (must be a valid python variable name)",
```
Comment (ptitzler) [inline elyra/pipeline/kfp/kfp_properties.py:76]:
```suggestion
            description="The name of the parameter. This must be a valid Python identifier and not a keyword.",
```

File: elyra/pipeline/kfp/kfp_properties.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            display_name="Parameter Name",
+            allowed_input_types=[PropertyInputType(base_type="str", placeholder="param_1")],
+            hidden=False,
+            required=True,
+            use_in_key=True,
+            pattern="^[a-zA-Z][a-zA-Z0-9_]*$",
+        ),
+        ListItemPropertyAttribute(
+            attribute_id="description",
+            description="A description for this parameter",
```
Comment (ptitzler) [inline elyra/pipeline/kfp/kfp_properties.py:86]: Not. to be consistent with help text format used for properties.

```suggestion
            description="A description for this parameter.",
```

File: elyra/pipeline/kfp/kfp_properties.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            display_name="Description",
+            allowed_input_types=[PropertyInputType(base_type="str")],
+            hidden=False,
+            required=False,
+            use_in_key=False,
+        ),
+        ListItemPropertyAttribute(
+            attribute_id="default_value",
+            description="A default value to assign to the parameter",
```
Comment (ptitzler) [inline elyra/pipeline/kfp/kfp_properties.py:95]:
```suggestion
            description="A default value for the parameter.",
```

File: elyra/pipeline/kfp/kfp_properties.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                KfpPropertyInputType(base_type="Float"),
+                KfpPropertyInputType(base_type="Bool"),
+            ],
+            hidden=True,
+            required=False,
+            use_in_key=False,
+        ),
+        ListItemPropertyAttribute(
+            attribute_id="required",
+            description="Whether a value is required for this parameter during pipeline submit/export",
```
Comment (ptitzler) [inline elyra/pipeline/kfp/kfp_properties.py:122]:
```suggestion
            description="Whether a value is required for this parameter during pipeline submit or export.",
```

**Comment by kiersten-stokes** [comment]
Note: the CodeQL warnings on this PR are 'the usual suspects' and have not been introduced by any changes made in this PR. (I think they may be triggered by the renaming of the `component_parameter.py` file to `properties.py`.)

**Comment by ptitzler** [comment]
> Note: the CodeQL warnings on this PR are 'the usual suspects' and have not been introduced by any changes made in this PR. (I think they may be triggered by the renaming of the `component_parameter.py` file to `properties.py`.)

I dismissed them.

**Review by ptitzler — APPROVED** [review]
LGTM. Thanks to you both for an awesome new feature!

**Pull request merged**